### DISCUSSION TITLE
Can it detach?

### DISCUSSION THREAD

**Discussion by kubouch** [wezterm/wezterm#712] [Q&A] [answered]
Hi,
I just discovered this project, it looks really cool! I'm definitely interested in trying it out.

I have a small question: can you detach from a running session and join later? Similar to tmux when you do `Ctrl-B+d` to detach, then `tmux a -t session` to attach back, without interrupting the running job.

Cheers,
Jakub

**Comment by wez** [comment, marked as answer]
If you explicitly enable the multiplexer server then you can configure a similar workflow: https://wezfurlong.org/wezterm/multiplexing.html

**Reply by kubouch** [reply]
Thanks! Apparently, I didn't read the docs carefully enough.

**Reply by losnappas** [reply]
Soo... How do I detach the window from inside the wezterm window?

If I launch wezterm from gnome-terminal with `wezterm connect unix`, then I can detach it by going ctrl-c from gnome-terminal, but how do I detach it from within? Closing the window always kills all the tabs. Am I making sense?

I saw "Hide" and "HideApplication", but neither seem to do anything on Ubuntu.

**Reply by wez** [reply]
wezterm doesn't have a first-class detach operation today... you can assign [QuitApplication](https://wezfurlong.org/wezterm/config/lua/keyassignment/QuitApplication.html) to quit the whole program without individually terminating the tabs/panes.

**Reply by losnappas** [reply]
By killing the gui process with `pkill wezterm-gui --signal SIGINT`, I can already get pretty close to a detach, but is this a bad approach? Unforeseen consequences?

Also, idk if I can add a keybind to send a custom command, however perhaps I can do that outside of wezterm, or learn to use a shell command to "detach", instead of the old `ctrl+b d`.

**Reply by wez** [reply]
It's "bad" in that it's a bad that wezterm doesn't have a better way.  It's potentially also bad if you setup custom keybindings that explicitly spawn a local program into the same gui process.

However, if you keep your multiplexer wezterm gui processes separate from the local ones, it's a workable solution.

You can set up a key binding to run a program without it spawning a tab.  Here's an example that does spawn a new tab:
Here's an example: https://wezfurlong.org/wezterm/config/lua/wezterm/on.html#custom-events
but you could change the Spawn line to use https://wezfurlong.org/wezterm/config/lua/wezterm/run_child_process.html if you wanted.

However, if all you're going to do is kill wezterm, you're probably better off just having a `QuitApplication` key assignment.

**Reply by losnappas** [reply]
> However, if all you're going to do is kill wezterm, you're probably better off just having a QuitApplication key assignment.

Oh, I misunderstood that one from the doc, thought it was gonna kill it and lose the terminal's history and whatnot, but it seems to do what I wanted indeed :+1: 

When I hit the X to close the window, it asks "Really kill this window..." which loses the history, the QuitApplication doc says "...killing all tabs.", but there seems to be a difference

**Reply by eugenesvk** [reply]
> assign [QuitApplication](https://wezfurlong.org/wezterm/config/lua/keyassignment/QuitApplication.html) to quit the whole program without individually terminating the tabs/panes.

How would you then quit the `wezterm-mux-server.exe` process on Windows after you've destroyed (not `QuitApplication`, but e.g. clicked on the `close window` button?

**Reply by wez** [reply]
I pushed https://github.com/wez/wezterm/commit/e9f1297b4559c2598c478c61fc84b3d351a85917 very recently, which improves things a bit here. We still don't have a formal dynamic detach/attach UX but at least it is easier to close a window without killing off everything in that window.

**Reply by wez** [reply]
> How would you then quit the `wezterm-mux-server.exe` process on Windows after you've destroyed (not `QuitApplication`, but e.g. clicked on the `close window` button?

If you don't want to gracefully terminate the panes it may be maintaining, you can use the task manager to kill it or:

`taskkill /im wezterm-mux-server.exe /f`

We don't have anything more integrated than that at the moment.

**Reply by eugenesvk** [reply]
> don't want to gracefully terminate

I like grace :) , but even when I `exit` a shell properly (or use `CloseCurrentPane`), the `wezterm-mux-server.exe` still remains in the background and blocks any new GUIs from maintaining a window (they flash and then just remain in the background)
Or are `exit`/`CloseCurrentPane` not graceful enough and there is something better?

> We still don't have a formal dynamic detach/attach UX

Ok, that's fine, just wanted to make sure I'm not missing anything obvious here

**Reply by wez** [reply]
> (they flash and then just remain in the background)

I'm assuming that you're using `connect_automatically` in this case?  I would recommend this instead: https://github.com/wez/wezterm/issues/1299#issuecomment-993770970

**Reply by eugenesvk** [reply]
> I'm assuming that you're using `connect_automatically` in this case? I would recommend this instead: [#1299 (comment)](https://github.com/wez/wezterm/issues/1299#issuecomment-993770970)

Yes, thanks a bunch, the solution is great even when I'm ungraceful ;) !
Now if only the window position stayed the same, but that's already mentioned in other issues...
